{"cells":[{"cell_type":"markdown","metadata":{"id":"DS5rVNdvHUUJ"},"source":["## Mount drive and import packages"]},{"cell_type":"code","execution_count":1,"metadata":{"executionInfo":{"elapsed":4034,"status":"ok","timestamp":1707470413181,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"b7z83HdQHQ3T"},"outputs":[{"name":"stderr","output_type":"stream","text":["/usr/local/lib/python3.10/dist-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n","  from .autonotebook import tqdm as notebook_tqdm\n"]}],"source":["from typing import List, Tuple, Dict, Optional, Set\n","from enum import Enum\n","import os\n","import pickle\n","import gc\n","import pandas as pd\n","from Bio import SeqIO\n","import matplotlib.pyplot as plt\n","import torch\n","from torch.utils.data import Dataset, DataLoader\n","import torch.nn.functional as F\n","import esm\n","from icecream import ic\n","import random\n","from collections import Counter\n","import requests\n","from urllib.parse import quote\n","from concurrent.futures import ThreadPoolExecutor, as_completed\n","from tqdm import tqdm\n","import sys\n","\n","sys.path.append('/workspace/protein_lm/tokenizer')\n","from tokenizer import EsmTokenizer, PTMTokenizer\n","\n","sys.path.append('/workspace/protein_lm/modeling/scripts')\n","from infer import PTMMamba"]},{"cell_type":"markdown","metadata":{"id":"NFRMeLSydWKL"},"source":["## Define token alphabet"]},{"cell_type":"code","execution_count":2,"metadata":{"executionInfo":{"elapsed":4,"status":"ok","timestamp":1707470413181,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"Kz_Kp-YqdXuY"},"outputs":[],"source":["tokens = [\n","            \"<cls>\",\n","            \"<pad>\",\n","            \"<eos>\",\n","            \"<unk>\",\n","            \".\",\n","            \"-\",\n","            \"<null_1>\",\n","            \"<mask>\",\n","            \"L\",\n","            \"A\",\n","            \"G\",\n","            \"V\",\n","            \"S\",\n","            \"E\",\n","            \"R\",\n","            \"T\",\n","            \"I\",\n","            \"D\",\n","            \"P\",\n","            \"K\",\n","            \"Q\",\n","            \"N\",\n","            \"F\",\n","            \"Y\",\n","            \"M\",\n","            \"H\",\n","            \"W\",\n","            \"C\",\n","            \"X\",\n","            \"B\",\n","            \"U\",\n","            \"Z\",\n","            \"O\",\n","            \"PTM\",\n","            \"<N-linked (GlcNAc...) asparagine>\",\n","            \"<Pyrrolidone carboxylic acid>\",\n","            \"<Phosphoserine>\",\n","            \"<Phosphothreonine>\",\n","            \"<N-acetylalanine>\",\n","            \"<N-acetylmethionine>\",\n","            \"<N6-acetyllysine>\",\n","            \"<Phosphotyrosine>\",\n","            \"<S-diacylglycerol cysteine>\",\n","            \"<N6-(pyridoxal phosphate)lysine>\",\n","            \"<N-acetylserine>\",\n","            \"<N6-carboxylysine>\",\n","            \"<N6-succinyllysine>\",\n","            \"<S-palmitoyl cysteine>\",\n","            \"<O-(pantetheine 4'-phosphoryl)serine>\",\n","            \"<Sulfotyrosine>\",\n","            \"<O-linked (GalNAc...) threonine>\",\n","            \"<Omega-N-methylarginine>\",\n","            \"<N-myristoyl glycine>\",\n","            \"<4-hydroxyproline>\",\n","            \"<Asymmetric dimethylarginine>\",\n","            \"<N5-methylglutamine>\",\n","            \"<4-aspartylphosphate>\",\n","            \"<S-geranylgeranyl cysteine>\",\n","            \"<4-carboxyglutamate>\",\n","        ]\n","token_to_index = {token: i for i, token in enumerate(tokens)}"]},{"cell_type":"markdown","metadata":{"id":"6Oo8kRDxHy-k"},"source":["## Get uniprot ids of proteins that are positive for druggability"]},{"cell_type":"code","execution_count":3,"metadata":{"executionInfo":{"elapsed":886,"status":"ok","timestamp":1707470414064,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"1Ek3BveHIW2R"},"outputs":[],"source":["positive_druggability_df_path = '/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/scraped_drugptm_table_with_uniprot_ids.csv'\n","positive_druggability_df = pd.read_csv(positive_druggability_df_path)\n","positive_druggability_ids = set(positive_druggability_df['UniProtID'])"]},{"cell_type":"markdown","metadata":{"id":"oGvg2gwqMave"},"source":["## Get uniprot ids of proteins that are positive for disease"]},{"cell_type":"code","execution_count":4,"metadata":{"executionInfo":{"elapsed":158,"status":"ok","timestamp":1707470414219,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"j5VjZaWbMcqV"},"outputs":[],"source":["positive_disease_df_path = '/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/scraped_snpptm_table_with_uniprot_ids.csv'\n","positive_disease_df = pd.read_csv(positive_disease_df_path)\n","positive_disease_ids = set(positive_disease_df['UniProtID'])"]},{"cell_type":"markdown","metadata":{"id":"dN-mDg5aMtu1"},"source":["## Get uniprot ids of proteins that are negative for druggability"]},{"cell_type":"code","execution_count":5,"metadata":{"executionInfo":{"elapsed":392,"status":"ok","timestamp":1707470414609,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"0vT-q0XAMr6_"},"outputs":[],"source":["negative_druggability_df_path = '/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/non_druggable_uniprot_ids.csv'\n","negative_druggability_df = pd.read_csv(negative_druggability_df_path)\n","negative_druggability_ids = set(negative_druggability_df['uniprot_id'])"]},{"cell_type":"markdown","metadata":{"id":"GgvmchNcO9rY"},"source":["## Get uniprot ids of proteins that are negative for disease"]},{"cell_type":"code","execution_count":6,"metadata":{"executionInfo":{"elapsed":390,"status":"ok","timestamp":1707470414996,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"GgZmBL_TOvR5"},"outputs":[],"source":["negative_disease_txt_path = '/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/non_disease_uniprot_ids.txt'\n","with open(negative_disease_txt_path, 'r') as file:\n","  negative_disease_ids = set([line.strip() for line in file])"]},{"cell_type":"code","execution_count":7,"metadata":{},"outputs":[],"source":["all_protein_ids = positive_druggability_ids | positive_disease_ids | negative_druggability_ids | negative_disease_ids"]},{"cell_type":"markdown","metadata":{"id":"hwqmicpjQNLy"},"source":["## Download all protein sequences"]},{"cell_type":"code","execution_count":8,"metadata":{"executionInfo":{"elapsed":4,"status":"ok","timestamp":1707470414996,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"M80Zz2-4Q5Dc"},"outputs":[],"source":["def download_protein_sequence(uniprot_id, file_path):\n","    base_url = \"https://rest.uniprot.org/uniprotkb/\"\n","    headers = {\"Accept\": \"text/plain\"}\n","    # Construct the query for a single UniProt ID\n","    url = f\"{base_url}/{uniprot_id}.fasta\"\n","    response = requests.get(url, headers=headers)\n","    if response.status_code == 200 and response.text:\n","        with open(file_path, 'a') as file:\n","            file.write(response.text)\n","            # print(f\"Downloaded sequence for UniProt ID: {uniprot_id}\")\n","    else:\n","        print(f\"Failed to download sequence for {uniprot_id}. Status code: {response.status_code}, Response: {response.text}\")\n","\n","def download_protein_sequences(uniprot_ids, file_path):\n","    for uniprot_id in tqdm(uniprot_ids):\n","        download_protein_sequence(uniprot_id, file_path)"]},{"cell_type":"code","execution_count":9,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":13158,"status":"ok","timestamp":1707470428150,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"SMjYc9LzT51E","outputId":"20b05c1f-277f-4996-912d-6c02ff624fc2"},"outputs":[{"name":"stderr","output_type":"stream","text":["ic| len(all_ids): 2466\n"]},{"data":{"text/plain":["2466"]},"execution_count":9,"metadata":{},"output_type":"execute_result"}],"source":["all_ids = list(positive_druggability_ids | positive_disease_ids | negative_druggability_ids | negative_disease_ids)\n","all_sequences_file_path = '/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/all_sequences.fasta'\n","if os.path.exists(all_sequences_file_path):\n","\tos.remove(all_sequences_file_path)\n","ic(len(all_ids))"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["download_protein_sequences(all_ids, all_sequences_file_path)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def fasta_to_dataframe(fasta_file):\n","    ids = []\n","    sequences = []\n","    for record in SeqIO.parse(fasta_file, \"fasta\"):\n","        uniprot_id = record.id.split('|')[1]\n","        sequence = str(record.seq)\n","        ids.append(uniprot_id)\n","        sequences.append(sequence)\n","    df = pd.DataFrame({\n","        'UniProt ID': ids,\n","        'Sequence': sequences\n","    })\n","    return df\n","\n","# wt_sequences_df = fasta_to_dataframe(all_sequences_file_path)\n","# ic(len(wt_sequences_df))\n","# wt_sequences_df.head()"]},{"cell_type":"markdown","metadata":{"id":"QN9BJ0nelsFP"},"source":["## Augment map uniprot ids to ptm_data.csv rows using their sequences"]},{"cell_type":"code","execution_count":11,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>AC_ID</th>\n","      <th>wt_seq</th>\n","      <th>ptm_seq</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>O60296</td>\n","      <td>MSQSQNAIFTSPTGEENLMNSNHRDSESITDVCSNEDLPEVELVSL...</td>\n","      <td>MSQSQNAIFTSPTGEENLMNSNHRDSESITDVCSNEDLPEVELVSL...</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>B5EXH1</td>\n","      <td>MTSRNYLLLTPGPLTTSRTVKEAMLFDSCTWDDDYNIGVVEQIRQQ...</td>\n","      <td>MTSRNYLLLTPGPLTTSRTVKEAMLFDSCTWDDDYNIGVVEQIRQQ...</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>Q9BXI9</td>\n","      <td>MQWLRVRESPGEATGHRVTMGTAALGPVWAALLLFLLMCEIPMVEL...</td>\n","      <td>MQWLRVRESPGEATGHRVTMGTAALGPVWAALLLFLLMCEIPMVEL...</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>Q6ZFW0</td>\n","      <td>MAASKGNAAAAACALVLVLLAVGAEAQGGGGGECVPQLNRLLACRA...</td>\n","      <td>MAASKGNAAAAACALVLVLLAVGAEAQGGGGGECVPQLNRLLACRA...</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Q3SYP2</td>\n","      <td>MLGITVLAAILACASSCGDPTFPPNLSARVVGGEDAVPNSWPWQVS...</td>\n","      <td>MLGITVLAAILACASSCGDPTFPP&lt;N-linked (GlcNAc...) ...</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["    AC_ID                                             wt_seq  \\\n","0  O60296  MSQSQNAIFTSPTGEENLMNSNHRDSESITDVCSNEDLPEVELVSL...   \n","1  B5EXH1  MTSRNYLLLTPGPLTTSRTVKEAMLFDSCTWDDDYNIGVVEQIRQQ...   \n","2  Q9BXI9  MQWLRVRESPGEATGHRVTMGTAALGPVWAALLLFLLMCEIPMVEL...   \n","3  Q6ZFW0  MAASKGNAAAAACALVLVLLAVGAEAQGGGGGECVPQLNRLLACRA...   \n","4  Q3SYP2  MLGITVLAAILACASSCGDPTFPPNLSARVVGGEDAVPNSWPWQVS...   \n","\n","                                             ptm_seq  \n","0  MSQSQNAIFTSPTGEENLMNSNHRDSESITDVCSNEDLPEVELVSL...  \n","1  MTSRNYLLLTPGPLTTSRTVKEAMLFDSCTWDDDYNIGVVEQIRQQ...  \n","2  MQWLRVRESPGEATGHRVTMGTAALGPVWAALLLFLLMCEIPMVEL...  \n","3  MAASKGNAAAAACALVLVLLAVGAEAQGGGGGECVPQLNRLLACRA...  \n","4  MLGITVLAAILACASSCGDPTFPP<N-linked (GlcNAc...) ...  "]},"execution_count":11,"metadata":{},"output_type":"execute_result"}],"source":["ptm_wt_df = pd.read_csv('/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/ptm_data.csv')\n","ptm_wt_df = ptm_wt_df[['AC_ID', 'wt_seq', 'ptm_seq']]\n","ptm_wt_df.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["ic| len(ptm_wt_df_seqs): 79707\n","ic| len(wt_sequences_df_seqs): 2461\n","ic| len(ptm_wt_df_seqs.intersection(wt_sequences_df_seqs)): 1682\n"]},{"data":{"text/plain":["1682"]},"execution_count":90,"metadata":{},"output_type":"execute_result"}],"source":["ptm_wt_df_seqs = set(ptm_wt_df['wt_seq'])\n","wt_sequences_df_seqs = set(wt_sequences_df['Sequence'])\n","\n","ic(len(ptm_wt_df_seqs))\n","ic(len(wt_sequences_df_seqs))\n","ic(len(ptm_wt_df_seqs.intersection(wt_sequences_df_seqs)))"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["ic| len(ac_ids): 79707\n","ic| len(all_protein_ids): 2466\n","ic| len(ac_ids.intersection(all_protein_ids)): 1619\n"]},{"data":{"text/plain":["1619"]},"execution_count":91,"metadata":{},"output_type":"execute_result"}],"source":["ac_ids = set(ptm_wt_df['AC_ID'])\n","ic(len(ac_ids))\n","ic(len(all_protein_ids))\n","ic(len(ac_ids.intersection(all_protein_ids)))"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"ename":"NameError","evalue":"name 'wt_sequences_df' is not defined","output_type":"error","traceback":["\u001b[0;31m---------------------------------------------------------------------------\u001b[0m","\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)","Cell \u001b[0;32mIn[9], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m seqs_df \u001b[38;5;241m=\u001b[39m df_joined \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mmerge(\u001b[43mwt_sequences_df\u001b[49m, ptm_wt_df, left_on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSequence\u001b[39m\u001b[38;5;124m'\u001b[39m, right_on\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwt_seq\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m      3\u001b[0m seqs_df\u001b[38;5;241m.\u001b[39mhead()\n","\u001b[0;31mNameError\u001b[0m: name 'wt_sequences_df' is not defined"]}],"source":["seqs_df = df_joined = pd.merge(wt_sequences_df, ptm_wt_df, left_on='Sequence', right_on='wt_seq')\n","\n","seqs_df.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>uniprot_id</th>\n","      <th>accession_id</th>\n","      <th>wt_seq</th>\n","      <th>ptm_seq</th>\n","      <th>one_hot_embedding_path</th>\n","      <th>one_hot_ptm_embedding_path</th>\n","      <th>esm_650m_embedding_path</th>\n","      <th>esm_3b_embedding_path</th>\n","      <th>ptm_mamba_embedding_path</th>\n","      <th>is_druggable</th>\n","      <th>is_disease</th>\n","      <th>is_part_of_druggability_dataset</th>\n","      <th>is_part_of_disease_dataset</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>P07107</td>\n","      <td>P07107</td>\n","      <td>MSQAEFDKAAEEVKHLKTKPADEEMLFIYSHYKQATVGDINTERPG...</td>\n","      <td>M&lt;N-acetylserine&gt;QAEFDKAAEEVKHL&lt;N6-succinyllys...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>O00590</td>\n","      <td>O00590</td>\n","      <td>MAATASPQPLATEDADSENSSFYYYDYLDEVAFMLCRKDAVVSFGK...</td>\n","      <td>MAATASPQPLATEDADSE&lt;N-linked (GlcNAc...) aspara...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>O75716</td>\n","      <td>O75716</td>\n","      <td>MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...</td>\n","      <td>MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>P23560</td>\n","      <td>Q5IS78</td>\n","      <td>MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...</td>\n","      <td>MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Q99471</td>\n","      <td>Q99471</td>\n","      <td>MAQSINITELNLPQLEMLKNQLDQEVEFLSTSIAQLKVVQTKYVEA...</td>\n","      <td>M&lt;N-acetylalanine&gt;QSINITELNLPQLEMLKNQLDQEVEFLS...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["  uniprot_id accession_id                                             wt_seq  \\\n","0     P07107       P07107  MSQAEFDKAAEEVKHLKTKPADEEMLFIYSHYKQATVGDINTERPG...   \n","1     O00590       O00590  MAATASPQPLATEDADSENSSFYYYDYLDEVAFMLCRKDAVVSFGK...   \n","2     O75716       O75716  MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...   \n","3     P23560       Q5IS78  MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...   \n","4     Q99471       Q99471  MAQSINITELNLPQLEMLKNQLDQEVEFLSTSIAQLKVVQTKYVEA...   \n","\n","                                             ptm_seq  \\\n","0  M<N-acetylserine>QAEFDKAAEEVKHL<N6-succinyllys...   \n","1  MAATASPQPLATEDADSE<N-linked (GlcNAc...) aspara...   \n","2  MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...   \n","3  MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...   \n","4  M<N-acetylalanine>QSINITELNLPQLEMLKNQLDQEVEFLS...   \n","\n","                              one_hot_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                          one_hot_ptm_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                             esm_650m_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                               esm_3b_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                            ptm_mamba_embedding_path  is_druggable  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...          True   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...          True   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","\n","   is_disease  is_part_of_druggability_dataset  is_part_of_disease_dataset  \n","0       False                             True                       False  \n","1        True                            False                        True  \n","2       False                             True                       False  \n","3        True                            False                        True  \n","4       False                             True                       False  "]},"execution_count":47,"metadata":{},"output_type":"execute_result"}],"source":["seqs_df.rename(columns={'UniProt ID': 'uniprot_id', \"AC_ID\": \"accession_id\"}, inplace=True)\n","seqs_df.drop('Sequence', axis=1, inplace=True)\n","\n","embeddings_dir = '/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/embeddings'\n","seqs_df['one_hot_embedding_path'] = seqs_df['uniprot_id'].apply(lambda x: f\"{embeddings_dir}/{x}_one_hot.pt\")\n","seqs_df['one_hot_ptm_embedding_path'] = seqs_df['uniprot_id'].apply(lambda x: f\"{embeddings_dir}/{x}_one_hot_ptm.pt\")\n","seqs_df['esm_650m_embedding_path'] = seqs_df['uniprot_id'].apply(lambda x: f\"{embeddings_dir}/{x}_esm_650m.pt\")\n","seqs_df['esm_3b_embedding_path'] = seqs_df['uniprot_id'].apply(lambda x: f\"{embeddings_dir}/{x}_esm_3b.pt\")\n","seqs_df['ptm_mamba_embedding_path'] = seqs_df['uniprot_id'].apply(lambda x: f\"{embeddings_dir}/{x}_ptm_mamba.pt\")\n","\n","seqs_df.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>uniprot_id</th>\n","      <th>accession_id</th>\n","      <th>wt_seq</th>\n","      <th>ptm_seq</th>\n","      <th>one_hot_embedding_path</th>\n","      <th>one_hot_ptm_embedding_path</th>\n","      <th>esm_650m_embedding_path</th>\n","      <th>esm_3b_embedding_path</th>\n","      <th>ptm_mamba_embedding_path</th>\n","      <th>is_druggable</th>\n","      <th>is_disease</th>\n","      <th>is_part_of_druggability_dataset</th>\n","      <th>is_part_of_disease_dataset</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>P07107</td>\n","      <td>P07107</td>\n","      <td>MSQAEFDKAAEEVKHLKTKPADEEMLFIYSHYKQATVGDINTERPG...</td>\n","      <td>M&lt;N-acetylserine&gt;QAEFDKAAEEVKHL&lt;N6-succinyllys...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>O00590</td>\n","      <td>O00590</td>\n","      <td>MAATASPQPLATEDADSENSSFYYYDYLDEVAFMLCRKDAVVSFGK...</td>\n","      <td>MAATASPQPLATEDADSE&lt;N-linked (GlcNAc...) aspara...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>O75716</td>\n","      <td>O75716</td>\n","      <td>MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...</td>\n","      <td>MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>P23560</td>\n","      <td>Q5IS78</td>\n","      <td>MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...</td>\n","      <td>MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Q99471</td>\n","      <td>Q99471</td>\n","      <td>MAQSINITELNLPQLEMLKNQLDQEVEFLSTSIAQLKVVQTKYVEA...</td>\n","      <td>M&lt;N-acetylalanine&gt;QSINITELNLPQLEMLKNQLDQEVEFLS...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["  uniprot_id accession_id                                             wt_seq  \\\n","0     P07107       P07107  MSQAEFDKAAEEVKHLKTKPADEEMLFIYSHYKQATVGDINTERPG...   \n","1     O00590       O00590  MAATASPQPLATEDADSENSSFYYYDYLDEVAFMLCRKDAVVSFGK...   \n","2     O75716       O75716  MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...   \n","3     P23560       Q5IS78  MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...   \n","4     Q99471       Q99471  MAQSINITELNLPQLEMLKNQLDQEVEFLSTSIAQLKVVQTKYVEA...   \n","\n","                                             ptm_seq  \\\n","0  M<N-acetylserine>QAEFDKAAEEVKHL<N6-succinyllys...   \n","1  MAATASPQPLATEDADSE<N-linked (GlcNAc...) aspara...   \n","2  MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...   \n","3  MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...   \n","4  M<N-acetylalanine>QSINITELNLPQLEMLKNQLDQEVEFLS...   \n","\n","                              one_hot_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                          one_hot_ptm_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                             esm_650m_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                               esm_3b_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                            ptm_mamba_embedding_path  is_druggable  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...          True   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...          True   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","\n","   is_disease  is_part_of_druggability_dataset  is_part_of_disease_dataset  \n","0       False                             True                       False  \n","1        True                            False                        True  \n","2       False                             True                       False  \n","3        True                            False                        True  \n","4       False                             True                       False  "]},"execution_count":126,"metadata":{},"output_type":"execute_result"}],"source":["seqs_df['is_druggable'] = seqs_df['uniprot_id'].apply(lambda x: x in positive_druggability_ids)\n","seqs_df['is_disease'] = seqs_df['uniprot_id'].apply(lambda x: x in positive_disease_ids)\n","\n","seqs_df['is_part_of_druggability_dataset'] = seqs_df['uniprot_id'].apply(lambda x: x in positive_druggability_ids or x in negative_druggability_ids)\n","seqs_df['is_part_of_disease_dataset'] = seqs_df['uniprot_id'].apply(lambda x: x in positive_disease_ids or x in negative_disease_ids)\n","\n","seqs_df.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["uniprot_id_to_index = {uniprot_id: i for i, uniprot_id in enumerate(seqs_df['uniprot_id'])}"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["seqs_df.to_csv('/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/seqs_df.csv', index=False)"]},{"cell_type":"code","execution_count":12,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>uniprot_id</th>\n","      <th>accession_id</th>\n","      <th>wt_seq</th>\n","      <th>ptm_seq</th>\n","      <th>one_hot_embedding_path</th>\n","      <th>one_hot_ptm_embedding_path</th>\n","      <th>esm_650m_embedding_path</th>\n","      <th>esm_3b_embedding_path</th>\n","      <th>ptm_mamba_embedding_path</th>\n","      <th>is_druggable</th>\n","      <th>is_disease</th>\n","      <th>is_part_of_druggability_dataset</th>\n","      <th>is_part_of_disease_dataset</th>\n","      <th>is_train</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>P07107</td>\n","      <td>P07107</td>\n","      <td>MSQAEFDKAAEEVKHLKTKPADEEMLFIYSHYKQATVGDINTERPG...</td>\n","      <td>M&lt;N-acetylserine&gt;QAEFDKAAEEVKHL&lt;N6-succinyllys...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>O00590</td>\n","      <td>O00590</td>\n","      <td>MAATASPQPLATEDADSENSSFYYYDYLDEVAFMLCRKDAVVSFGK...</td>\n","      <td>MAATASPQPLATEDADSE&lt;N-linked (GlcNAc...) aspara...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>O75716</td>\n","      <td>O75716</td>\n","      <td>MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...</td>\n","      <td>MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>P23560</td>\n","      <td>Q5IS78</td>\n","      <td>MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...</td>\n","      <td>MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Q99471</td>\n","      <td>Q99471</td>\n","      <td>MAQSINITELNLPQLEMLKNQLDQEVEFLSTSIAQLKVVQTKYVEA...</td>\n","      <td>M&lt;N-acetylalanine&gt;QSINITELNLPQLEMLKNQLDQEVEFLS...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["  uniprot_id accession_id                                             wt_seq  \\\n","0     P07107       P07107  MSQAEFDKAAEEVKHLKTKPADEEMLFIYSHYKQATVGDINTERPG...   \n","1     O00590       O00590  MAATASPQPLATEDADSENSSFYYYDYLDEVAFMLCRKDAVVSFGK...   \n","2     O75716       O75716  MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...   \n","3     P23560       Q5IS78  MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...   \n","4     Q99471       Q99471  MAQSINITELNLPQLEMLKNQLDQEVEFLSTSIAQLKVVQTKYVEA...   \n","\n","                                             ptm_seq  \\\n","0  M<N-acetylserine>QAEFDKAAEEVKHL<N6-succinyllys...   \n","1  MAATASPQPLATEDADSE<N-linked (GlcNAc...) aspara...   \n","2  MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...   \n","3  MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...   \n","4  M<N-acetylalanine>QSINITELNLPQLEMLKNQLDQEVEFLS...   \n","\n","                              one_hot_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                          one_hot_ptm_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                             esm_650m_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                               esm_3b_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                            ptm_mamba_embedding_path  is_druggable  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...          True   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...          True   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","\n","   is_disease  is_part_of_druggability_dataset  is_part_of_disease_dataset  \\\n","0       False                             True                       False   \n","1        True                            False                        True   \n","2       False                             True                       False   \n","3        True                            False                        True   \n","4       False                             True                       False   \n","\n","   is_train  \n","0      True  \n","1      True  \n","2      True  \n","3      True  \n","4      True  "]},"execution_count":12,"metadata":{},"output_type":"execute_result"}],"source":["seqs_df = pd.read_csv('/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/seqs_df.csv')\n","\n","seqs_df.head()"]},{"cell_type":"markdown","metadata":{},"source":["## Divide datasets into train and test sets using mmmseqs2"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["drugptm_seqs_df = seqs_df[seqs_df['is_part_of_druggability_dataset']]\n","snpptm_seqs_df = seqs_df[seqs_df['is_part_of_disease_dataset']]"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["# iteeate over the rows of drugptm_seqs_df and copy all wt_seqs to a fasta file\n","drugptm_seqs_file_path = '/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/clustering/drugptm_sequences.fasta'\n","\n","with open(drugptm_seqs_file_path, 'a') as file:\n","\tfor index, row in drugptm_seqs_df.iterrows():\n","\t\tfile.write(f\">{row['uniprot_id']}\\n{row['wt_seq']}\\n\")\n","\n","\n","disease_seqs_file_path = '/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/clustering/disease_sequences.fasta'\n","with open(disease_seqs_file_path, 'a') as file:\n","\tfor index, row in snpptm_seqs_df.iterrows():\n","\t\tfile.write(f\">{row['uniprot_id']}\\n{row['wt_seq']}\\n\")"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/clustering\n"]},{"name":"stderr","output_type":"stream","text":["/usr/local/lib/python3.10/dist-packages/IPython/core/magics/osm.py:417: UserWarning: using dhist requires you to install the `pickleshare` library.\n","  self.shell.db['dhist'] = compress_dhist(dhist)[-100:]\n"]}],"source":["%cd /workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/clustering"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[33mdrugptmProteinDB exists and will be overwritten\n","\u001b[39mcreatedb /workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/clustering/drugptm_sequences.fasta drugptmProteinDB \n","\n","MMseqs Version:       \tabab073141f0bba788aed068b376de99c4c3519a\n","Database type         \t0\n","Shuffle input database\ttrue\n","Createdb mode         \t0\n","Write lookup file     \t1\n","Offset of numeric ids \t0\n","Compressed            \t0\n","Verbosity             \t3\n","\n","Converting sequences\n","[607] 0s 3ms\n","Time for merging to drugptmProteinDB_h: 0h 0m 0s 1ms\n","Time for merging to drugptmProteinDB: 0h 0m 0s 2ms\n","Database type: Aminoacid\n","Time for processing: 0h 0m 0s 10ms\n"]}],"source":["!mmseqs createdb /workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/clustering/drugptm_sequences.fasta drugptmProteinDB"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["cluster drugptmProteinDB drugptmClusters tmp --min-seq-id 0.3 -c 0.5 --cov-mode 1 --cluster-mode 0 \n","\n","MMseqs Version:                     \tabab073141f0bba788aed068b376de99c4c3519a\n","Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n","Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n","Sensitivity                         \t4\n","k-mer length                        \t0\n","Target search mode                  \t0\n","k-score                             \tseq:2147483647,prof:2147483647\n","Alphabet size                       \taa:21,nucl:5\n","Max sequence length                 \t65535\n","Max results per query               \t20\n","Split database                      \t0\n","Split mode                          \t2\n","Split memory limit                  \t0\n","Coverage threshold                  \t0.5\n","Coverage mode                       \t1\n","Compositional bias                  \t1\n","Compositional bias                  \t1\n","Diagonal scoring                    \ttrue\n","Exact k-mer matching                \t0\n","Mask residues                       \t1\n","Mask residues probability           \t0.9\n","Mask lower case residues            \t0\n","Minimum diagonal score              \t15\n","Selected taxa                       \t\n","Include identical seq. id.          \tfalse\n","Spaced k-mers                       \t1\n","Preload mode                        \t0\n","Pseudo count a                      \tsubstitution:1.100,context:1.400\n","Pseudo count b                      \tsubstitution:4.100,context:5.800\n","Spaced k-mer pattern                \t\n","Local temporary path                \t\n","Threads                             \t256\n","Compressed                          \t0\n","Verbosity                           \t3\n","Add backtrace                       \tfalse\n","Alignment mode                      \t3\n","Alignment mode                      \t0\n","Allow wrapped scoring               \tfalse\n","E-value threshold                   \t0.001\n","Seq. id. threshold                  \t0.3\n","Min alignment length                \t0\n","Seq. id. mode                       \t0\n","Alternative alignments              \t0\n","Max reject                          \t2147483647\n","Max accept                          \t2147483647\n","Score bias                          \t0\n","Realign hits                        \tfalse\n","Realign score bias                  \t-0.2\n","Realign max seqs                    \t2147483647\n","Correlation score weight            \t0\n","Gap open cost                       \taa:11,nucl:5\n","Gap extension cost                  \taa:1,nucl:2\n","Zdrop                               \t40\n","Rescore mode                        \t0\n","Remove hits by seq. id. and coverage\tfalse\n","Sort results                        \t0\n","Cluster mode                        \t0\n","Max connected component depth       \t1000\n","Similarity type                     \t2\n","Weight file name                    \t\n","Cluster Weight threshold            \t0.9\n","Single step clustering              \tfalse\n","Cascaded clustering steps           \t3\n","Cluster reassign                    \tfalse\n","Remove temporary files              \tfalse\n","Force restart with latest tmp       \tfalse\n","MPI runner                          \t\n","k-mers per sequence                 \t21\n","Scale k-mers per sequence           \taa:0.000,nucl:0.200\n","Adjust k-mer length                 \tfalse\n","Shift hash                          \t67\n","Include only extendable             \tfalse\n","Skip repeating k-mers               \tfalse\n","\n","Set cluster sensitivity to -s 5.000000\n","\u001b[33mCombining cluster mode 0 in combination with coverage mode 1 can produce wrong results.\n","Please use --cov-mode 2\n","\u001b[39mSet cluster iterations to 3\n","drugptmClusters.dbtype exists already!\n"]}],"source":["# !mmseqs cluster drugptmProteinDB drugptmClusters tmp --min-seq-id 0.3 -c 0.5 --cov-mode 1\n","\n","!mmseqs cluster drugptmProteinDB drugptmClusters tmp --min-seq-id 0.3 -c 0.5 --cov-mode 1 --cluster-mode 0"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["createtsv drugptmProteinDB drugptmProteinDB drugptmClusters drugptmClusters.tsv \n","\n","MMseqs Version:                 \tabab073141f0bba788aed068b376de99c4c3519a\n","First sequence as representative\tfalse\n","Target column                   \t1\n","Add full header                 \tfalse\n","Sequence source                 \t0\n","Database output                 \tfalse\n","Threads                         \t256\n","Compressed                      \t0\n","Verbosity                       \t3\n","\n","Time for merging to drugptmClusters.tsv: 0h 0m 0s 11ms\n","Time for processing: 0h 0m 0s 126ms\n"]}],"source":["!mmseqs createtsv drugptmProteinDB drugptmProteinDB drugptmClusters drugptmClusters.tsv"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def read_clusters_dict_from_tsv(tsv_path):\n","    clusters = {}\n","    with open(tsv_path) as f:\n","        for line in f:\n","            parts = line.strip().split(\"\\t\")\n","            cluster_id = parts[0]\n","            sequence_id = parts[1]\n","            if cluster_id not in clusters:\n","                clusters[cluster_id] = set()\n","            clusters[cluster_id].add(sequence_id)\n","    return clusters\n","\n","disease_tsv_path = \"/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/clustering/drugptmClusters.tsv\"\n","drugptm_clusters = read_clusters_dict_from_tsv(disease_tsv_path)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["def random_split_clusters(clusters):\n","    random.seed(42)\n","\n","    cluster_ids = list(clusters.keys())\n","    random.shuffle(cluster_ids)\n","\n","    num_test = int(len(cluster_ids) * 0.2)\n","    test_clusters = cluster_ids[:num_test]\n","    train_clusters = cluster_ids[num_test:]\n","\n","    test_ids = {seq_id for cluster_id in test_clusters for seq_id in clusters[cluster_id]}\n","    train_ids = {seq_id for cluster_id in train_clusters for seq_id in clusters[cluster_id]}\n","\n","    return train_ids, test_ids\n","\n","drugptm_train_protein_ids, drugptm_test_protein_ids = random_split_clusters(drugptm_clusters)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["\u001b[33mdiseaseProteinDB exists and will be overwritten\n","\u001b[39mcreatedb /workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/clustering/disease_sequences.fasta diseaseProteinDB \n","\n","MMseqs Version:       \tabab073141f0bba788aed068b376de99c4c3519a\n","Database type         \t0\n","Shuffle input database\ttrue\n","Createdb mode         \t0\n","Write lookup file     \t1\n","Offset of numeric ids \t0\n","Compressed            \t0\n","Verbosity             \t3\n","\n","Converting sequences\n","[1011] 0s 4ms\n","Time for merging to diseaseProteinDB_h: 0h 0m 0s 2ms\n","Time for merging to diseaseProteinDB: 0h 0m 0s 4ms\n","Database type: Aminoacid\n","Time for processing: 0h 0m 0s 16ms\n"]}],"source":["!mmseqs createdb /workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/clustering/disease_sequences.fasta diseaseProteinDB"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["cluster diseaseProteinDB diseaseClusters tmp --min-seq-id 0.3 -c 0.5 --cov-mode 1 --cluster-mode 0 \n","\n","MMseqs Version:                     \tabab073141f0bba788aed068b376de99c4c3519a\n","Substitution matrix                 \taa:blosum62.out,nucl:nucleotide.out\n","Seed substitution matrix            \taa:VTML80.out,nucl:nucleotide.out\n","Sensitivity                         \t4\n","k-mer length                        \t0\n","Target search mode                  \t0\n","k-score                             \tseq:2147483647,prof:2147483647\n","Alphabet size                       \taa:21,nucl:5\n","Max sequence length                 \t65535\n","Max results per query               \t20\n","Split database                      \t0\n","Split mode                          \t2\n","Split memory limit                  \t0\n","Coverage threshold                  \t0.5\n","Coverage mode                       \t1\n","Compositional bias                  \t1\n","Compositional bias                  \t1\n","Diagonal scoring                    \ttrue\n","Exact k-mer matching                \t0\n","Mask residues                       \t1\n","Mask residues probability           \t0.9\n","Mask lower case residues            \t0\n","Minimum diagonal score              \t15\n","Selected taxa                       \t\n","Include identical seq. id.          \tfalse\n","Spaced k-mers                       \t1\n","Preload mode                        \t0\n","Pseudo count a                      \tsubstitution:1.100,context:1.400\n","Pseudo count b                      \tsubstitution:4.100,context:5.800\n","Spaced k-mer pattern                \t\n","Local temporary path                \t\n","Threads                             \t256\n","Compressed                          \t0\n","Verbosity                           \t3\n","Add backtrace                       \tfalse\n","Alignment mode                      \t3\n","Alignment mode                      \t0\n","Allow wrapped scoring               \tfalse\n","E-value threshold                   \t0.001\n","Seq. id. threshold                  \t0.3\n","Min alignment length                \t0\n","Seq. id. mode                       \t0\n","Alternative alignments              \t0\n","Max reject                          \t2147483647\n","Max accept                          \t2147483647\n","Score bias                          \t0\n","Realign hits                        \tfalse\n","Realign score bias                  \t-0.2\n","Realign max seqs                    \t2147483647\n","Correlation score weight            \t0\n","Gap open cost                       \taa:11,nucl:5\n","Gap extension cost                  \taa:1,nucl:2\n","Zdrop                               \t40\n","Rescore mode                        \t0\n","Remove hits by seq. id. and coverage\tfalse\n","Sort results                        \t0\n","Cluster mode                        \t0\n","Max connected component depth       \t1000\n","Similarity type                     \t2\n","Weight file name                    \t\n","Cluster Weight threshold            \t0.9\n","Single step clustering              \tfalse\n","Cascaded clustering steps           \t3\n","Cluster reassign                    \tfalse\n","Remove temporary files              \tfalse\n","Force restart with latest tmp       \tfalse\n","MPI runner                          \t\n","k-mers per sequence                 \t21\n","Scale k-mers per sequence           \taa:0.000,nucl:0.200\n","Adjust k-mer length                 \tfalse\n","Shift hash                          \t67\n","Include only extendable             \tfalse\n","Skip repeating k-mers               \tfalse\n","\n","Set cluster sensitivity to -s 5.000000\n","\u001b[33mCombining cluster mode 0 in combination with coverage mode 1 can produce wrong results.\n","Please use --cov-mode 2\n","\u001b[39mSet cluster iterations to 3\n","diseaseClusters.dbtype exists already!\n"]}],"source":["!mmseqs cluster diseaseProteinDB diseaseClusters tmp --min-seq-id 0.3 -c 0.5 --cov-mode 1 --cluster-mode 0\n","\n","# !mmseqs cluster diseaseProteinDB diseaseClusters tmp --min-seq-id 0.3 -c 0.5 --cov-mode 1"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["createtsv diseaseProteinDB diseaseProteinDB diseaseClusters diseaseClusters.tsv \n","\n","MMseqs Version:                 \tabab073141f0bba788aed068b376de99c4c3519a\n","First sequence as representative\tfalse\n","Target column                   \t1\n","Add full header                 \tfalse\n","Sequence source                 \t0\n","Database output                 \tfalse\n","Threads                         \t256\n","Compressed                      \t0\n","Verbosity                       \t3\n","\n","Time for merging to diseaseClusters.tsv: 0h 0m 0s 12ms\n","Time for processing: 0h 0m 0s 116ms\n"]}],"source":["!mmseqs createtsv diseaseProteinDB diseaseProteinDB diseaseClusters diseaseClusters.tsv"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["disease_tsv_path = \"/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/clustering/diseaseClusters.tsv\"\n","disease_clusters = read_clusters_dict_from_tsv(disease_tsv_path)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["disease_train_protein_ids, disease_test_protein_ids = random_split_clusters(disease_clusters)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["ic| len(drugptm_train_protein_ids): 552\n","ic| len(drugptm_test_protein_ids): 144\n","ic| len(disease_train_protein_ids): 809\n","ic| len(disease_test_protein_ids): 221\n","ic| len(set(drugptm_train_protein_ids).intersection(drugptm_test_protein_ids)): 0\n","ic| len(set(disease_train_protein_ids).intersection(disease_test_protein_ids)): 0\n","ic| list(drugptm_train_protein_ids)[:5]: ['P21457', 'P11142', 'P00875', 'Q04760', 'P07814']\n","ic| list(drugptm_test_protein_ids)[:5]: ['P80386', 'P31939', 'Q9X2A5', 'Q8TD30', 'P35475']\n","ic| list(disease_train_protein_ids)[:5]: ['P13866', 'Q8N5U6', 'Q8IV20', 'Q7Z7F7', 'P78346']\n","ic| list(disease_test_protein_ids)[:5]: ['Q969V6', 'P62955', 'Q3MIW9', 'O95613', 'Q9NTG1']\n"]},{"data":{"text/plain":["['Q969V6', 'P62955', 'Q3MIW9', 'O95613', 'Q9NTG1']"]},"execution_count":85,"metadata":{},"output_type":"execute_result"}],"source":["ic(len(drugptm_train_protein_ids))\n","ic(len(drugptm_test_protein_ids))\n","ic(len(disease_train_protein_ids))\n","ic(len(disease_test_protein_ids))\n","\n","ic(len(set(drugptm_train_protein_ids).intersection(drugptm_test_protein_ids)))\n","ic(len(set(disease_train_protein_ids).intersection(disease_test_protein_ids)))\n","\n","ic(list(drugptm_train_protein_ids)[:5])\n","ic(list(drugptm_test_protein_ids)[:5])\n","ic(list(disease_train_protein_ids)[:5])\n","ic(list(disease_test_protein_ids)[:5])"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>uniprot_id</th>\n","      <th>accession_id</th>\n","      <th>wt_seq</th>\n","      <th>ptm_seq</th>\n","      <th>one_hot_embedding_path</th>\n","      <th>one_hot_ptm_embedding_path</th>\n","      <th>esm_650m_embedding_path</th>\n","      <th>esm_3b_embedding_path</th>\n","      <th>ptm_mamba_embedding_path</th>\n","      <th>is_druggable</th>\n","      <th>is_disease</th>\n","      <th>is_part_of_druggability_dataset</th>\n","      <th>is_part_of_disease_dataset</th>\n","      <th>is_train</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>0</th>\n","      <td>P07107</td>\n","      <td>P07107</td>\n","      <td>MSQAEFDKAAEEVKHLKTKPADEEMLFIYSHYKQATVGDINTERPG...</td>\n","      <td>M&lt;N-acetylserine&gt;QAEFDKAAEEVKHL&lt;N6-succinyllys...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>1</th>\n","      <td>O00590</td>\n","      <td>O00590</td>\n","      <td>MAATASPQPLATEDADSENSSFYYYDYLDEVAFMLCRKDAVVSFGK...</td>\n","      <td>MAATASPQPLATEDADSE&lt;N-linked (GlcNAc...) aspara...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>2</th>\n","      <td>O75716</td>\n","      <td>O75716</td>\n","      <td>MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...</td>\n","      <td>MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>3</th>\n","      <td>P23560</td>\n","      <td>Q5IS78</td>\n","      <td>MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...</td>\n","      <td>MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>True</td>\n","    </tr>\n","    <tr>\n","      <th>4</th>\n","      <td>Q99471</td>\n","      <td>Q99471</td>\n","      <td>MAQSINITELNLPQLEMLKNQLDQEVEFLSTSIAQLKVVQTKYVEA...</td>\n","      <td>M&lt;N-acetylalanine&gt;QSINITELNLPQLEMLKNQLDQEVEFLS...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>/workspace/protein_lm/evaluation/nonptm_vs_ptm...</td>\n","      <td>False</td>\n","      <td>False</td>\n","      <td>True</td>\n","      <td>False</td>\n","      <td>True</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["  uniprot_id accession_id                                             wt_seq  \\\n","0     P07107       P07107  MSQAEFDKAAEEVKHLKTKPADEEMLFIYSHYKQATVGDINTERPG...   \n","1     O00590       O00590  MAATASPQPLATEDADSENSSFYYYDYLDEVAFMLCRKDAVVSFGK...   \n","2     O75716       O75716  MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...   \n","3     P23560       Q5IS78  MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...   \n","4     Q99471       Q99471  MAQSINITELNLPQLEMLKNQLDQEVEFLSTSIAQLKVVQTKYVEA...   \n","\n","                                             ptm_seq  \\\n","0  M<N-acetylserine>QAEFDKAAEEVKHL<N6-succinyllys...   \n","1  MAATASPQPLATEDADSE<N-linked (GlcNAc...) aspara...   \n","2  MGHALCVCSRGTVIIDNKRYLFIQKLGEGGFSYVDLVEGLHDGHFY...   \n","3  MTILFLTMVISYFGCMKAAPMKEANIRGQGGLAYPGVRTHGTLESV...   \n","4  M<N-acetylalanine>QSINITELNLPQLEMLKNQLDQEVEFLS...   \n","\n","                              one_hot_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                          one_hot_ptm_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                             esm_650m_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                               esm_3b_embedding_path  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...   \n","\n","                            ptm_mamba_embedding_path  is_druggable  \\\n","0  /workspace/protein_lm/evaluation/nonptm_vs_ptm...          True   \n","1  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","2  /workspace/protein_lm/evaluation/nonptm_vs_ptm...          True   \n","3  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","4  /workspace/protein_lm/evaluation/nonptm_vs_ptm...         False   \n","\n","   is_disease  is_part_of_druggability_dataset  is_part_of_disease_dataset  \\\n","0       False                             True                       False   \n","1        True                            False                        True   \n","2       False                             True                       False   \n","3        True                            False                        True   \n","4       False                             True                       False   \n","\n","   is_train  \n","0      True  \n","1      True  \n","2      True  \n","3      True  \n","4      True  "]},"execution_count":86,"metadata":{},"output_type":"execute_result"}],"source":["seqs_df['is_train'] = seqs_df['uniprot_id'].apply(lambda x: x in drugptm_train_protein_ids or x in disease_train_protein_ids)\n","\n","seqs_df.head()"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["seqs_df.to_csv('/workspace/protein_lm/evaluation/nonptm_vs_ptm_classification/data/seqs_df.csv', index=False)"]},{"cell_type":"markdown","metadata":{"id":"I2g5zYq6bXWI"},"source":["## Create one-hot embeddings"]},{"cell_type":"code","execution_count":24,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":362328,"status":"ok","timestamp":1707471722589,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"PnT0ywyQdMRZ","outputId":"02bd728a-7b4f-4ae1-bfe8-a6761923f659"},"outputs":[{"name":"stderr","output_type":"stream","text":["1686it [00:00, 2647.85it/s]\n"]}],"source":["tokenizer = PTMTokenizer()\n","\n","for i, row in tqdm(seqs_df.iterrows()):\n","    uniprot_id = row['uniprot_id']\n","    wt_seq = row['wt_seq']\n","    one_hot_embedding_path = row['one_hot_embedding_path']\n","    one_hot_vector = tokenizer(wt_seq, return_tensor=True)\n","    padded_one_hot_vector = F.pad(one_hot_vector, (0, 7000 - one_hot_vector.shape[0]), \"constant\", 0)\n","    torch.save(padded_one_hot_vector, one_hot_embedding_path)"]},{"cell_type":"markdown","metadata":{"id":"jYDtTZF5e8OZ"},"source":["## Create one-hot PTM embeddings"]},{"cell_type":"code","execution_count":25,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":27433,"status":"ok","timestamp":1707471749999,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"8burqAJZe-4X","outputId":"0d7d409d-d4b6-4c9a-b4ed-ac472a701862"},"outputs":[{"name":"stderr","output_type":"stream","text":["1686it [00:00, 2615.06it/s]\n"]}],"source":["tokenizer = PTMTokenizer()\n","\n","for i, row in tqdm(seqs_df.iterrows()):\n","    uniprot_id = row['uniprot_id']\n","    ptm_seq = row['ptm_seq']\n","    one_hot_ptm_embedding_path = row['one_hot_ptm_embedding_path']\n","    one_hot_ptm_vector = tokenizer(ptm_seq, return_tensor=True)\n","    padded_one_hot_ptm_vector = F.pad(one_hot_ptm_vector, (0, 7000 - one_hot_ptm_vector.shape[0]), \"constant\", 0)\n","    torch.save(padded_one_hot_ptm_vector, one_hot_ptm_embedding_path)"]},{"cell_type":"markdown","metadata":{"id":"i6uyi8klndA-"},"source":["## Create ESM 650M Embeddings"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":22572,"status":"ok","timestamp":1707471772553,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"lWtQapvTBJ46","outputId":"ad539f46-b428-4173-a170-dd0d0f4e4871"},"outputs":[{"name":"stderr","output_type":"stream","text":["ic| device: device(type='cuda', index=0)\n","ic| torch.cuda.is_available(): True\n"]},{"data":{"text/plain":["True"]},"execution_count":57,"metadata":{},"output_type":"execute_result"}],"source":["esm_650m_model, esm_650m_alphabet = esm.pretrained.esm2_t33_650M_UR50D()\n","esm_650m_batch_converter = esm_650m_alphabet.get_batch_converter()\n","esm_650m_model.eval()\n","device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n","esm_650m_model.to(device)\n","ic(device)\n","ic(torch.cuda.is_available())"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":4,"status":"ok","timestamp":1707471772553,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"7iWERxXhAaQD"},"outputs":[],"source":["class ESMProteinDataset(Dataset):\n","    def __init__(self,\n","                batch_converter,\n","                device,\n","                seqs_df):\n","        self.batch_converter = batch_converter\n","        self.device = device\n","        self.df = seqs_df\n","\n","    def __len__(self):\n","        return len(self.df)\n","\n","    def __getitem__(self, idx):\n","        row = self.df.iloc[idx]\n","        uniprot_id = row['uniprot_id']\n","        wt_seq = row['wt_seq']\n","        _, _, batch_tokens = self.batch_converter([(uniprot_id, wt_seq)])\n","        return uniprot_id, batch_tokens[0].to(device)\n","\n","torch.manual_seed(0)\n","esm_650m_dataset = ESMProteinDataset(esm_650m_batch_converter, device, seqs_df)\n","esm_650m_loader = DataLoader(esm_650m_dataset, batch_size=1, shuffle=True)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":283180,"status":"ok","timestamp":1707472055730,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"MaV_krlVAaF_","outputId":"feabdf50-94da-4692-91a4-54194635774f"},"outputs":[{"name":"stderr","output_type":"stream","text":["100%|| 1686/1686 [01:16<00:00, 21.91it/s]\n"]}],"source":["for protein_ids, batch_tokens in tqdm(esm_650m_loader):\n","  with torch.no_grad():\n","    try:\n","      results = esm_650m_model(batch_tokens, repr_layers=[33])\n","    except RuntimeError as e:\n","      ic(e)\n","      ic(protein_ids)\n","      ic(protein_ids)\n","      continue\n","  token_embeddings = results[\"representations\"][33]\n","  for i, protein_id in enumerate(protein_ids):\n","    index = uniprot_id_to_index[protein_id]\n","    row = seqs_df.iloc[index]\n","    esm_650m_embedding_path = row['esm_650m_embedding_path']\n","    averaged_embedding = token_embeddings[i].mean(dim=0)\n","    with open(esm_650m_embedding_path, 'wb') as f:\n","      torch.save(averaged_embedding, f)"]},{"cell_type":"markdown","metadata":{"id":"nXAGVdr8HRPN"},"source":["## Create ESM 3B Embeddings"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":62540,"status":"ok","timestamp":1707472118264,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"X_0ODoVgAZ5t","outputId":"fe9992b6-47fd-44d0-f513-c1c878ed6663"},"outputs":[{"name":"stderr","output_type":"stream","text":["ic| device: device(type='cuda', index=0)\n"]},{"data":{"text/plain":["device(type='cuda', index=0)"]},"execution_count":60,"metadata":{},"output_type":"execute_result"}],"source":["esm_650m_model.cpu()\n","gc.collect()\n","torch.cuda.empty_cache()\n","\n","esm_3b_model, esm_3b_alphabet = esm.pretrained.esm2_t36_3B_UR50D()\n","esm_3b_batch_converter = esm_3b_alphabet.get_batch_converter()\n","esm_3b_model.eval()\n","esm_3b_model.to(device)\n","ic(device)"]},{"cell_type":"code","execution_count":null,"metadata":{"executionInfo":{"elapsed":3,"status":"ok","timestamp":1707472118264,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"O8h_Zx4aAZ_u"},"outputs":[],"source":["torch.manual_seed(0)\n","esm_3b_dataset = ESMProteinDataset(esm_3b_batch_converter, device, seqs_df)\n","esm_3b_loader = DataLoader(esm_3b_dataset, batch_size=1, shuffle=True)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":795353,"status":"ok","timestamp":1707472913614,"user":{"displayName":"Benjamin Schussheim","userId":"06913257060266462744"},"user_tz":300},"id":"lTafcymLFieG","outputId":"9988da96-06bb-4760-c1bb-630c9f0eb21e"},"outputs":[{"name":"stderr","output_type":"stream","text":["100%|| 1686/1686 [02:20<00:00, 11.99it/s]\n"]}],"source":["for protein_ids, batch_tokens in tqdm(esm_3b_loader):\n","  with torch.no_grad():\n","    try:\n","      results = esm_3b_model(batch_tokens, repr_layers=[33])\n","    except RuntimeError as e:\n","      ic(e)\n","      ic(protein_ids)\n","      ic(protein_ids)\n","      continue\n","  token_embeddings = results[\"representations\"][33]\n","  for i, protein_id in enumerate(protein_ids):\n","    index = uniprot_id_to_index[protein_id]\n","    row = seqs_df.iloc[index]\n","    esm_3b_embedding_path = row['esm_3b_embedding_path']\n","    averaged_embedding = token_embeddings[i].mean(dim=0)\n","    with open(esm_3b_embedding_path, 'wb') as f:\n","      torch.save(averaged_embedding, f)"]},{"cell_type":"markdown","metadata":{},"source":["## Create PTMMamba Embeddings"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":["class PTMProteinDataset(Dataset):\n","    def __init__(self,\n","                device,\n","                seqs_df):\n","        self.device = device\n","        self.df = seqs_df\n","\n","    def __len__(self):\n","        return len(self.df)\n","\n","    def __getitem__(self, idx):\n","        row = self.df.iloc[idx]\n","        uniprot_id = row['uniprot_id']\n","        ptm_seq = row['ptm_seq']\n","        return uniprot_id, ptm_seq\n","\n","torch.manual_seed(0)\n","device = torch.device(\"cuda:0\" if torch.cuda.is_available() else \"cpu\")\n","ptm_mamba_dataset = PTMProteinDataset(device, seqs_df)\n","ptm_mamba_loader = DataLoader(ptm_mamba_dataset, batch_size=1, shuffle=False)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stdout","output_type":"stream","text":["<All keys matched successfully>\n"]}],"source":["model_checkpoint_path = '/workspace/ckpt/bi_mamba-esm-ptm_token_input/best.ckpt'\n","mamba = PTMMamba(ckpt_path=model_checkpoint_path, device=device)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[{"name":"stderr","output_type":"stream","text":["100%|| 1686/1686 [02:14<00:00, 12.58it/s]\n"]}],"source":["ptm_tokenizer = PTMTokenizer()\n","\n","for protein_id, ptm_seq in tqdm(ptm_mamba_loader):\n","    ptm_seq = ptm_seq[0]\n","    protein_id = protein_id[0]\n","    index = uniprot_id_to_index[protein_id]\n","    ptm_mamba_embedding_path = seqs_df.iloc[index]['ptm_mamba_embedding_path']\n","    tokenized_output = ptm_tokenizer(ptm_seq, return_tensor=True)\n","    with torch.no_grad():\n","        try:\n","            output = mamba(ptm_seq)\n","            embedding = output.hidden_states.squeeze(dim=0)\n","            averaged_embedding = embedding.mean(dim=0)\n","            with open(ptm_mamba_embedding_path, 'wb') as f:\n","                torch.save(averaged_embedding, f)\n","        except RuntimeError as e:\n","            ic(e)\n","            ic(protein_id)"]},{"cell_type":"code","execution_count":null,"metadata":{},"outputs":[],"source":[]}],"metadata":{"accelerator":"GPU","colab":{"authorship_tag":"ABX9TyN6bLcWYQ2/hb4dovgpU8IH","gpuType":"A100","machine_shape":"hm","provenance":[]},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.10.12"}},"nbformat":4,"nbformat_minor":0}
